File: src/main/java/dev/ainterr/minecraft/blockshuffle/PlayerList.java
package dev.ainterr.minecraft.blockshuffle;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.Player;

import java.util.HashMap;
import java.util.Random;
import java.util.Set;


public class PlayerList {
    public static final int RETURN_SUCCESS = 0;
    public static final int RETURN_FAILURE = 1;

    public static final int STATUS_SUCCESS = 0;
    public static final int STATUS_FAILURE = 1;
    public static final int STATUS_WAITING = -1;

    private final HashMap<Player, Material> blocks = new HashMap<>();
    private final HashMap<Player, Integer> status = new HashMap<>();

    private void initializePlayer(Player player) {
        this.blocks.put(player, null);
        this.status.put(player, STATUS_WAITING);
    }

    public boolean existsPlayer(Player player) {
        return this.status.containsKey(player);
    }

    public void addPlayer(Player player) {
        if (this.existsPlayer(player)) {
            return;
        }

        this.initializePlayer(player);
    }

    public int newBlock(Player player, Material selection) {
        if (!this.existsPlayer(player)) {
            return RETURN_FAILURE;
        }

        this.initializePlayer(player);

        if (!player.isOnline()) {
            return RETURN_FAILURE;
        }

        Random r = new Random();

        while (selection == null) {
            selection = Material.values()[r.nextInt(Material.values().length)];

            if (!selection.isBlock() || BlockSetsKt.getDefaultBlacklist().contains(selection)) {
                selection = null;
            }
        }

        this.blocks.put(player, selection);
        this.status.put(player, STATUS_FAILURE);

        return RETURN_SUCCESS;
    }

    public int newBlock(Player player) {
        return this.newBlock(player, null);
    }

    private boolean checkBlock(Player player) {
        Location location = player.getLocation();

        Material target = this.blocks.get(player);

         Material block = location.getBlock().getType();

        if (block == target) {
            return true;
        }

        block = location.clone().subtract(0, 1, 0).getBlock().getType();

        return block == target;
    }

    public boolean isBlockFound(Player player) {
        if (!this.existsPlayer(player)) {
            return false;
        }

        if (!player.isOnline()) {
            this.initializePlayer(player);
            return false;
        }

        if (this.status.get(player) == STATUS_WAITING) {
            return false;
        } else if (this.status.get(player) == STATUS_SUCCESS) {
            return true;
        }

        boolean found = this.checkBlock(player);

        if (found) {
            this.status.put(player, STATUS_SUCCESS);
        }

        return found;
    }

    public Material getBlockMaterial(Player player) {
        if (!this.existsPlayer(player)) {
            return null;
        }

        return this.blocks.get(player);
    }

    public String getBlock(Player player) {
        if (!this.existsPlayer(player)) {
            return "";
        }

        return this.blocks.get(player).toString().replace('_', ' ');
    }

    public int getStatus(Player player) {
        if (!this.existsPlayer(player)) {
            return STATUS_WAITING;
        }

        return this.status.get(player);
    }

    public int getTotalStatus() {
        int totalStatus = STATUS_SUCCESS;

        for (Player player : this.getPlayers()) {
            totalStatus += this.getStatus(player);
        }

        return totalStatus;
    }

    public Set<Player> getPlayers() {
        return this.status.keySet();
    }
}

File: src/main/java/dev/ainterr/minecraft/blockshuffle/PlayerList.kt
package dev.ainterr.minecraft.blockshuffle

import org.bukkit.Material
import org.bukkit.block.BlockFace
import org.bukkit.entity.Player
import java.util.*

class PlayerList {
    private val blocks = HashMap<Player, Material?>()
    private val status = HashMap<Player, Int>()

    val players: Set<Player>
        get() = status.keys

    private fun initializePlayer(player: Player) {
        blocks[player] = null
        status[player] = STATUS_WAITING
    }

    private fun existsPlayer(player: Player): Boolean {
        return status.containsKey(player)
    }

    fun addPlayer(player: Player) {
        if (existsPlayer(player)) {
            return
        }
        initializePlayer(player)
    }

    fun newBlock(player: Player, selection: Material?) {
        var newSelection = selection
        if (!existsPlayer(player)) {
            return
        }
        initializePlayer(player)
        if (!player.isOnline) {
            return
        }
        if (newSelection == null) {
            val r = Random()
            val targetBlocks = SimpleTargetBlocks
            newSelection = targetBlocks[r.nextInt(targetBlocks.size)]
        }
        blocks[player] = newSelection
        status[player] = STATUS_FAILURE
    }

    private fun checkBlock(player: Player): Boolean {
        val playerLocation = player.location
        val target = blocks[player]
        val block = playerLocation.block.type
        if (block == target) {
            return true
        }
        val materialUnderPlayer = playerLocation.block.getRelative(BlockFace.DOWN).type
        return materialUnderPlayer == target
    }

    fun isBlockFound(player: Player): Boolean {
        if (!existsPlayer(player)) {
            return false
        }
        if (!player.isOnline) {
            initializePlayer(player)
            return false
        }
        if (status[player] == STATUS_WAITING) {
            return false
        } else if (status[player] == STATUS_SUCCESS) {
            return true
        }
        val found = checkBlock(player)
        if (found) {
            status[player] = STATUS_SUCCESS
        }
        return found
    }

    fun getBlockMaterial(player: Player): Material? {
        return if (!existsPlayer(player)) {
            null
        } else blocks[player]
    }

    fun getBlock(player: Player): String {
        return if (!existsPlayer(player)) {
            ""
        } else blocks[player].toString().replace('_', ' ')
    }

    fun getStatus(player: Player): Int {
        return if (!existsPlayer(player)) {
            STATUS_WAITING
        } else status[player]!!
    }

    val totalStatus: Int
        get() {
            var totalStatus = STATUS_SUCCESS
            for (player in players) {
                totalStatus += getStatus(player)
            }
            return totalStatus
        }

    companion object {
        const val STATUS_SUCCESS = 0
        const val STATUS_FAILURE = 1
        const val STATUS_WAITING = -1
    }
}